### PR TITLE
Get testing

### PR DESCRIPTION
**Pull request by alexandercbooth**
## Status
OK, we can now issue menu events with spectron 😄.

A few issues currently:
- no graceful exit -- for each test, when we quit the notebook with `app.stop()`, we get our dialog about making sure so you don't lose unsaved data. I haven't figured out how to dismiss this (press the "Yes" button) yet, so for now I am just using the electron api to destroy the browser window.  Also, since we have async function calls jest is unhappy about this ungraceful exit, so I added in a `--forceExit` arg.
- I think at the moment, only node based notebooks are possible to test. We would have to do additional installs on the CI image to get the right (python, IPython kernel etc.).  We absolutely want to do testing against different kernels, but the moment I think it's good to get the process right with the bundled kernel.

## Questions
- the tests are really just running the menu options, not actually asserting anything. What should the actual test assertion strategy be here?
- more long term, what would we ideally like to test across the whole app?

### CODE REVIEW

**Comment by rgbkrk** [comment]
I think using the node.js kernel is perfectly 👌 and eases our path to testing this. We should be able to do everything, messages wise. 

As for closing the window, I wonder if there's a way for us to trigger the save with a filename. Thinking about what the menu actions respond to, I think that might be best. Then we can even do a jest snapshot with that notebook! 🤤

**Comment by alexandercbooth** [comment]
>As for closing the window, I wonder if there's a way for us to trigger the save with a filename. Thinking about what the menu actions respond to, I think that might be best. Then we can even do a jest snapshot with that notebook! 🤤

That is an :100: _awesome_ idea! I think I see the way forward 🙂.

Will update tonight, thanks @rgbkrk!

**Comment by alexandercbooth** [comment]
### Update
Chuffed with this now 😁. 

- Gracefully handling close by saving 😄.
 
Had a working solution where I copied notebooks manually to avoid rewriting on save and then figured out that `.send("menu:save-as", "thingy.ipynb")` just _worked_. 

Super clever idea of snapshotting these, @rgbkrk!

Sometimes I almost wonder if the tools available to to us are _too_ good 🤔. It's a good time to be coding 🙂.

**Comment by rgbkrk** [comment]
Awesommmmeeee, this is what I've always dreamed of.

File: __tests__/integration.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
+const Application = require("spectron").Application;
+const assert = require("assert");
+const path = require("path");
+const fs = require("fs");
+
+const appPath = "nteract/applications/desktop/dist/squashfs-root/nteract";
+const cleanNotebookPath = "notebooks/node-example.ipynb";
```
Comment (rgbkrk) [inline __tests__/integration.js]: Nice.

File: __tests__/integration.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    await app.browserWindow.send("menu:save-as", "notebooks/clear-all.ipynb");
+    // Setting the path for the starting notebook for the next test. Inelegant for sure, but it works ¯\_(ツ)_/¯
+    nextTestNb = cleanNotebookPath;
+  });
+
+  it("Open clean notebook example, runs all cells", async () => {
+    await delay(12000);
+    await app.client.windowByIndex(0);
+    await app.browserWindow.send("menu:run-all");
+    await app.browserWindow.send("menu:save-as", "notebooks/run-all.ipynb");
```
Comment (rgbkrk) [inline __tests__/integration.js]: Hmmmm.... We probably need to watch for when the notebook is "done" executing. We'd probably have to hook into the store to see when it's idle I assume (?)

**Comment by rgbkrk** [comment]
I'm going to merge this gets us on the right path. We should probably keep discussing approaches in here too. How do we ensure errors propagate?

File: __tests__/integration.js (hunk 2 above)
Comment (alexandercbooth) [reply __tests__/integration.js]: That's a good idea! -- I wasn't sure the best approach to get away from using `--forceExit` 😅

**Comment by alexandercbooth** [comment]
>I'm going to merge this gets us on the right path. We should probably keep discussing approaches in here too. How do we ensure errors propagate?

That sounds good 👍
Definitely need to think more about error propagation 🤔

File: __tests__/integration.js (hunk 2 above)
Comment (rgbkrk) [reply __tests__/integration.js]: I was thinking mostly about how `run-all` the command will finish before the cells are really done running, which means the save as will be missing the outputs.

Comment (alexandercbooth) [reply __tests__/integration.js]: Ohhhh gotcha! I hadn't thought of that. Will try to incorporate in the future 👍

**Pull request merged**